File: lib/ui/overlays.dart
import 'package:flutter/material.dart';

import '../engine/game_controller.dart';
import '../render/game_painter.dart';
import 'hud.dart';

class MenuOverlay extends StatelessWidget {
  const MenuOverlay({super.key, required this.game});

  final GameController game;

  @override
  Widget build(BuildContext context) {
    return _Scrim(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          const Text('THE ZERO',
              style: TextStyle(
                color: Colors.white,
                fontSize: 34,
                fontWeight: FontWeight.w900,
                letterSpacing: 4,
              )),
          const Text('FOAM CHALLENGE',
              style: TextStyle(
                color: Palette.accent,
                fontSize: 30,
                fontWeight: FontWeight.w900,
                letterSpacing: 2,
              )),
          const SizedBox(height: 18),
          Text(
            'Pour down the tilted wall for a laminar,\n'
            'foam-free flow. Fill to '
            '${(GameController.targetLiquid * 100).round()}% before the timer '
            'runs out —\nand never let the glass overflow.',
            textAlign: TextAlign.center,
            style: TextStyle(
              color: Colors.white.withValues(alpha: 0.7),
              fontSize: 14,
              height: 1.5,
            ),
          ),
          const SizedBox(height: 24),
          _ModeSwitch(game: game),
          const SizedBox(height: 24),
          PillButton(label: 'POUR', onTap: game.startRound),
        ],
      ),
    );
  }
}

class _ModeSwitch extends StatelessWidget {
  const _ModeSwitch({required this.game});

  final GameController game;

  @override
  Widget build(BuildContext context) {
    Widget option(ControlMode mode, String label, bool enabled) {
      final active = game.controlMode == mode;
      return Opacity(
        opacity: enabled ? 1 : 0.35,
        child: GestureDetector(
          onTap: enabled ? () => game.setControlMode(mode) : null,
          child: Container(
            padding: const EdgeInsets.symmetric(horizontal: 18, vertical: 10),
            decoration: BoxDecoration(
              color: active
                  ? Palette.accent.withValues(alpha: 0.2)
                  : Colors.white.withValues(alpha: 0.05),
              borderRadius: BorderRadius.circular(20),
              border: Border.all(
                color: active ? Palette.accent : Colors.white24,
              ),
            ),
            child: Text(
              label,
              style: TextStyle(
                color: active ? Palette.accent : Colors.white70,
                fontWeight: FontWeight.w800,
                fontSize: 12,
                letterSpacing: 1.2,
              ),
            ),
          ),
        ),
      );
    }

    return Column(
      children: [
        Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            option(ControlMode.touch, 'TOUCH', true),
            const SizedBox(width: 10),
            option(ControlMode.tilt, 'TILT + TOUCH', game.sensorAvailable),
          ],
        ),
        if (!game.sensorAvailable)
          Padding(
            padding: const EdgeInsets.only(top: 8),
            child: Text(
              'No motion sensor detected — touch controls only',
              style: TextStyle(
                color: Colors.white.withValues(alpha: 0.4),
                fontSize: 11,
              ),
            ),
          ),
      ],
    );
  }
}

class ResultOverlay extends StatelessWidget {
  const ResultOverlay({super.key, required this.game, required this.success});

  final GameController game;
  final bool success;

  @override
  Widget build(BuildContext context) {
    final r = game.lastResult;
    return _Scrim(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Text(
            success ? 'NICE POUR!' : 'ROUND OVER',
            style: TextStyle(
              color: success ? Palette.accent : Palette.danger,
              fontSize: 34,
              fontWeight: FontWeight.w900,
              letterSpacing: 3,
            ),
          ),
          if (r != null) ...[
            const SizedBox(height: 6),
            Text(r.reason,
                style: const TextStyle(
                  color: Colors.white70,
                  fontSize: 14,
                  letterSpacing: 2,
                  fontWeight: FontWeight.w700,
                )),
            const SizedBox(height: 20),
            Text('${r.score}',
                style: const TextStyle(
                  color: Colors.white,
                  fontSize: 56,
                  fontWeight: FontWeight.w900,
                )),
            const SizedBox(height: 12),
            _row('Liquid', '${(r.liquid * 100).toStringAsFixed(1)}%'),
            _row('Foam', '${(r.foam * 100).toStringAsFixed(1)}%'),
            _row('Time', '${r.timeUsed.toStringAsFixed(1)}s'),
            _row('Multiplier', '×${r.multiplier.toStringAsFixed(2)}'),
            _row('Best', '${game.bestScore}'),
          ],
          const SizedBox(height: 26),
          Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              PillButton(label: 'MENU', onTap: game.toMenu, primary: false),
              const SizedBox(width: 12),
              PillButton(label: 'RETRY', onTap: game.startRound),
            ],
          ),
        ],
      ),
    );
  }

  static Widget _row(String label, String value) => Padding(
        padding: const EdgeInsets.symmetric(vertical: 3),
        child: SizedBox(
          width: 220,
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Text(label,
                  style: const TextStyle(color: Colors.white54, fontSize: 14)),
              Text(value,
                  style: const TextStyle(
                    color: Colors.white,
                    fontSize: 14,
                    fontWeight: FontWeight.w800,
                  )),
            ],
          ),
        ),
      );
}

class _Scrim extends StatelessWidget {
  const _Scrim({required this.child});

  final Widget child;

  @override
  Widget build(BuildContext context) {
    return Container(
      color: const Color(0xCC050B15),
      alignment: Alignment.center,
      child: SingleChildScrollView(
        padding: const EdgeInsets.symmetric(horizontal: 28, vertical: 40),
        child: child,
      ),
    );
  }
}

File: ZeroFoamChallenge/Sources/Views/Overlays.swift
import SwiftUI

struct MenuOverlay: View {
    @ObservedObject var game: GameController

    var body: some View {
        Scrim {
            VStack(spacing: 0) {
                Text("THE ZERO")
                    .font(.system(size: 34, weight: .black)).kerning(4)
                    .foregroundColor(.white)
                Text("FOAM CHALLENGE")
                    .font(.system(size: 30, weight: .black)).kerning(2)
                    .foregroundColor(Palette.accent)
                Text("""
                     Pour down the tilted wall for a laminar, foam-free flow. \
                     Fill to \(Int(GameController.targetLiquid * 100))% before the \
                     timer runs out — and never let the glass overflow.
                     """)
                    .multilineTextAlignment(.center)
                    .font(.system(size: 14))
                    .lineSpacing(4)
                    .foregroundColor(.white.opacity(0.7))
                    .padding(.top, 18)
                ModeSwitch(game: game).padding(.top, 24)
                PillButton(label: "POUR") { game.startRound() }.padding(.top, 24)
            }
        }
    }
}

struct ModeSwitch: View {
    @ObservedObject var game: GameController

    var body: some View {
        VStack(spacing: 8) {
            HStack(spacing: 10) {
                option(.touch, "TOUCH", enabled: true)
                option(.tilt, "TILT + TOUCH", enabled: game.sensorAvailable)
            }
            if !game.sensorAvailable {
                Text("No motion sensor detected — touch controls only")
                    .font(.system(size: 11))
                    .foregroundColor(.white.opacity(0.4))
            }
        }
    }

    private func option(_ mode: ControlMode, _ label: String, enabled: Bool) -> some View {
        let active = game.controlMode == mode
        return Text(label)
            .font(.system(size: 12, weight: .heavy)).kerning(1.2)
            .foregroundColor(active ? Palette.accent : .white.opacity(0.7))
            .padding(.horizontal, 18)
            .padding(.vertical, 10)
            .background(
                Capsule().fill(active ? Palette.accent.opacity(0.2) : Color.white.opacity(0.05))
                    .overlay(Capsule().stroke(active ? Palette.accent : Color.white.opacity(0.24),
                                              lineWidth: 1))
            )
            .opacity(enabled ? 1 : 0.35)
            .onTapGesture { if enabled { game.setControlMode(mode) } }
    }
}

struct ResultOverlay: View {
    @ObservedObject var game: GameController
    let success: Bool

    var body: some View {
        Scrim {
            VStack(spacing: 0) {
                Text(success ? "NICE POUR!" : "ROUND OVER")
                    .font(.system(size: 34, weight: .black)).kerning(3)
                    .foregroundColor(success ? Palette.accent : Palette.danger)
                if let r = game.lastResult {
                    Text(r.reason)
                        .font(.system(size: 14, weight: .bold)).kerning(2)
                        .foregroundColor(.white.opacity(0.7))
                        .padding(.top, 6)
                    Text("\(r.score)")
                        .font(.system(size: 56, weight: .black))
                        .foregroundColor(.white)
                        .padding(.top, 20)
                    VStack(spacing: 6) {
                        row("Liquid", String(format: "%.1f%%", r.liquid * 100))
                        row("Foam", String(format: "%.1f%%", r.foam * 100))
                        row("Time", String(format: "%.1fs", r.timeUsed))
                        row("Multiplier", String(format: "×%.2f", r.multiplier))
                        row("Best", "\(game.bestScore)")
                    }
                    .padding(.top, 12)
                }
                HStack(spacing: 12) {
                    PillButton(label: "MENU", primary: false) { game.toMenu() }
                    PillButton(label: "RETRY") { game.startRound() }
                }
                .padding(.top, 26)
            }
        }
    }

    private func row(_ label: String, _ value: String) -> some View {
        HStack {
            Text(label).font(.system(size: 14)).foregroundColor(.white.opacity(0.55))
            Spacer()
            Text(value).font(.system(size: 14, weight: .heavy)).foregroundColor(.white)
        }
        .frame(width: 220)
    }
}

struct Scrim<Content: View>: View {
    @ViewBuilder var content: Content

    var body: some View {
        ZStack {
            Color(red: 0.02, green: 0.04, blue: 0.08).opacity(0.8).ignoresSafeArea()
            ScrollView {
                content.padding(.horizontal, 28).padding(.vertical, 40)
            }
        }
    }
}
